### PR TITLE
Issue #43: Use a single reaper goroutine

### PR DESCRIPTION
**Pull request by maniksurtani**
See #43

### CODE REVIEW

File: reaper.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+func (r *reaper) checkForNewWatchers() {
+	t := time.NewTicker(r.cfg.BucketWatcherFrequency)
+
+	for range t.C {
+		if !r.checkIsRunning() {
+			break
+		}
+
+		for w := range r.newWatchers {
```
Comment (maniksurtani) [inline reaper.go]: That's a good point. Should we just allow it to block? If we do then we can lose the timer (and corresponding cfg). Just depends on how expensive it is for a goroutine to block on a channel.

Comment (steved) [inline reaper.go]: Are you worried about the cost of locking and adding a new watcher? That should happen more frequently than the reaper itself, so I don't think it'd hurt too much.

File: reaper.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	}
+}
+
+// tooIdle returns true if a bucket has been idle for longer than its maxIdle
+func (w *watcher) tooIdle(now time.Time) bool {
+	if w.activityDetected() {
+		w.lastActivity = now
+		return false
+	}
+	if now.Sub(w.lastActivity) > w.maxIdle {
```
Comment (maniksurtani) [inline reaper.go]: Yup

File: server.go [hunk 3]
```diff
@@ -196,7 +198,7 @@ func (s *server) createBucketContainer(newConfig *pb.ServiceConfig) {
 	s.bucketFactory.Init(newConfig)
 
 	s.cfgs = newConfig
-	s.bucketContainer = NewBucketContainer(s.cfgs, s.bucketFactory, s)
+	s.bucketContainer = NewBucketContainer(s.cfgs, s.bucketFactory, s, s.reaperConfig)
 	s.Unlock()
```
Comment (maniksurtani) [inline server.go]: Yup

File: bucket.go [hunk 4]
```diff
@@ -275,21 +238,22 @@ func (bc *bucketContainer) countDynamicBuckets(namespace string) int32 {
...
 }
 
-func (bc *bucketContainer) createNewNamedBucketFromCfg(namespace, bucketName string, ns *namespace, bCfg *pbconfig.BucketConfig, dyn bool) *expirableBucket {
+func (bc *bucketContainer) createNewNamedBucketFromCfg(namespace, bucketName string, ns *namespace, bCfg *pbconfig.BucketConfig, dyn bool) Bucket {
 	bc.n.Emit(events.NewBucketCreatedEvent(namespace, bucketName, dyn))
-	bucket := bc.newExpirableBucket(namespace, bucketName, bCfg, dyn)
+	var bucket Bucket
+	bucket = bc.bf.NewBucket(namespace, bucketName, bCfg, dyn)
+
+	if bucket != nil {
```
Comment (maniksurtani) [inline bucket.go]: Fixed

File: reaper.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+	defer r.Unlock()
+	r.watchers[w.identifier] = w
+	w.lastActivity = time.Now()
+}
+
+// checkExpirations checks all watches registered with the reaper, and destroys idle buckets, updating the reaper
+// accordingly. Returns the duration after which it should run again.
+func (r *reaper) checkExpirations(bc *bucketContainer) time.Duration {
+	r.Lock()
+	defer r.Unlock()
```
Comment (steved) [inline reaper.go]: Do you need to lock here (and in addNewWatcher)?

File: reaper.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+			// Exit loop if stopped
+			if !r.checkIsRunning() {
+				break
+			}
+
+			r.addNewWatcher(w)
+
+		case <-ticker.C:
+			// Exit loop if stopped
+			if !r.checkIsRunning() {
```
Comment (steved) [inline reaper.go]: `running` could also be a channel, removing the need for these two checks

**Comment by maniksurtani** [comment]
Even nicer, yeah!

File: reaper.go (hunk 5 above)
Comment (maniksurtani) [inline reaper.go]: Not anymore, since it's a single goroutine now instead of 2. Good catch.

**Comment by maniksurtani** [comment]
@steved PTAL again - I also updated a few places where channels weren't specified as being read-only or write-only.

File: reaper.go [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+	for {
+		select {
+		case w, ok := <-newWatchers:
+			if ok {
+				r.addNewWatcher(w)
+			} else {
+				// newWatchers closed; stop the reaper.
+				r.newWatchers = nil
+				r.watchers = nil
+				break
```
Comment (steved) [inline reaper.go]: You'll want to return, since `break` only get you out of the select.

**Comment by steved** [comment]
One last comment, otherwise :+1:

File: reaper.go (hunk 7 above)
Comment (maniksurtani) [inline reaper.go]: Good catch!

**Pull request merged**